### PR TITLE
Feature/add 2d beam

### PR DESCRIPTION
**Pull request by KlausBSautter**
Hey I added the 2 d version of the cr beam. I did it in a new file to avoid too much code being nested and I could go for an easier rotation strategy. 
There are additional tests and I did several non-linear benchmarks which all work pretty nicely

### CODE REVIEW

**Review by loumalouomega — APPROVED** [review]
OK, looks fine

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp [hunk 1]
```diff
@@ -0,0 +1,887 @@
...
+// System includes
+
+// External includes
+
+// Project includes
+#include "custom_elements/cr_beam_element_2D2N.hpp"
+#include "structural_mechanics_application_variables.h"
+#include "includes/define.h"
+
+#define Pi 3.141592653589793238462643383279502884197169399375105820974944592308
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp]: Please use `Globals::Pi`

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp [hunk 2]
```diff
@@ -0,0 +1,887 @@
...
+	}
+
+	bounded_matrix<double,CrBeamElement2D2N::msElementSize,
+	CrBeamElement2D2N::msLocalSize> CrBeamElement2D2N::CalculateTransformationS() 
+	{
+		KRATOS_TRY;
+		const double L = this->CalculateCurrentLength();
+		bounded_matrix<double,msElementSize,msLocalSize> S = ZeroMatrix(msElementSize, msLocalSize);
+		S(0, 0) = -1.00;
+		S(1, 2) = 2.00 / L;
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp:559]: what if `L` is very small? => add check (maybe even in `CalculateCurrentLength`-function)

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp [hunk 3]
```diff
@@ -0,0 +1,887 @@
...
+		if (this->GetProperties().Has(AREA_EFFECTIVE_Y)) {
+			Ay = GetProperties()[AREA_EFFECTIVE_Y];
+		}
+
+		const double Psi = this->CalculatePsi(Iz, Ay);
+
+		// element material stiffness matrix
+		bounded_matrix<double,msLocalSize,msLocalSize> Kd = ZeroMatrix(msLocalSize, msLocalSize);
+
+		Kd(0, 0) = E*A / L;
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp:592]: same

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp [hunk 4]
```diff
@@ -0,0 +1,887 @@
...
+
+
+	bounded_matrix<double,CrBeamElement2D2N::msElementSize,CrBeamElement2D2N::msElementSize>
+	 CrBeamElement2D2N::CreateElementStiffnessMatrix_Kr()
+	 {
+		KRATOS_TRY	
+		// element properties
+		const double L = this->CalculateCurrentLength();
+		const double N = this->DeformationForces[0];
+		const double Q = (-2.00 / L) * this->DeformationForces[2];
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp:624]: same

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp [hunk 5]
```diff
@@ -0,0 +1,887 @@
...
+		return qe;
+		KRATOS_CATCH("")
+	 }
+
+
+
+	 double CrBeamElement2D2N::Modulus2Pi(double A)
+	 {
+		KRATOS_TRY;	
+		const int B = A / (2.00 * Pi);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp]: globals::Pi

**Review by philbucher — CHANGES_REQUESTED** [review]
Just some very minor changes
Also I will upload a small fix, see #1083

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp [hunk 6]
```diff
@@ -0,0 +1,892 @@
...
+// Project includes
+#include "custom_elements/cr_beam_element_2D2N.hpp"
+#include "structural_mechanics_application_variables.h"
+#include "includes/define.h"
+
+
+namespace Kratos
+{
+
+	CrBeamElement2D2N::CrBeamElement2D2N(IndexType NewId,
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp:26]: honestly i don't love this mechanism.

you are storing a bool in each element for this purpose, and you are using a non-standard constructor. Would it be much more complex to use derivation (for example deriving the linear beam from the nonlinear beam, or the contrary)?

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.hpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+		bounded_vector<double,msLocalSize> CalculateInternalStresses_DeformationModes();
+		bounded_vector<double,msElementSize> ReturnElementForces_Local();
+
+	private:
+
+		bool mIsLinearElement = false;
+		bounded_vector<double,msLocalSize> DeformationForces = ZeroVector(msLocalSize);
+		Vector F_int_global = ZeroVector(msElementSize);
+		Matrix K_master = ZeroMatrix(msElementSize,msElementSize);
+		CrBeamElement2D2N() {};
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.hpp:157]: define it as

CrBeamElement2D2N() = delete;

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.hpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+		bounded_vector<double,msLocalSize> CalculateDeformationParameters();
+		bounded_vector<double,msLocalSize> CalculateInternalStresses_DeformationModes();
+		bounded_vector<double,msElementSize> ReturnElementForces_Local();
+
+	private:
+
+		bool mIsLinearElement = false;
+		bounded_vector<double,msLocalSize> DeformationForces = ZeroVector(msLocalSize);
+		Vector F_int_global = ZeroVector(msElementSize);
+		Matrix K_master = ZeroMatrix(msElementSize,msElementSize);
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.hpp:156]: do you really need to store this?

**Review by RiccardoRossi — COMMENTED** [review]
minor comments

**Review by philbucher — APPROVED** [review]

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.hpp (hunk 7 above)
Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.hpp:157]: I need this for the serializer

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.hpp (hunk 8 above)
Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.hpp:156]: Yes I need the internal forces which are converged from the last iteration of the last step in my RHS

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp (hunk 6 above)
Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_2D2N.cpp:26]: I will merge this branch now and discuss with @AndreasWinterstein  and @philbucher  about this possibility, as I can then also change the other elements

**Pull request merged**